askup/tests/test_functional/test_general.py

```diff
@@ -1443,3 +1459,169 @@ def test_user_has_no_questions(self):
+        self.assertContains(response, 'SUBSCRIBE')
+        self.assertNotContains(response, 'UNSUBSCRIBE')
+
+        self.change_qset_subscription(self, qset_id, 'subscribe')  # Subscribe current user to the selected qset
+
+        response = self.get_my_subscriptions(self)
+        self.assertContains(response, 'UNSUBSCRIBE')
+
+        self.change_qset_subscription(self, qset_id, 'unsubscribe')  # Unsubscribe current user from the selected qset
+
+        response = self.get_my_subscriptions(self)
+        self.assertNotContains(response, 'UNSUBSCRIBE')
+
+    @client_user('student01', 'student01')
+    def test_subscription_success(self):
+        """
+        Test successful subscription.
+        """
+        qset_id = 4  # Qset 1-1 from the mockups
+        # Subscribe current user to the selected qset
+        response = self.change_qset_subscription(self, qset_id, 'subscribe').json()
+        self.assertEqual(response['result'], 'success')
+        self.assertEqual(
+            response['url'],
+            reverse(
+                'askup:qset_subscription',
+                kwargs={'qset_id': qset_id, 'subscribe': 'unsubscribe'}
+            )
+        )
+
+        # Unsubscribe current user to the selected qset
+        response = self.change_qset_subscription(self, qset_id, 'unsubscribe').json()
+        self.assertEqual(response['result'], 'success')
+        self.assertEqual(
+            response['url'],
+            reverse(
+                'askup:qset_subscription',
+                kwargs={'qset_id': qset_id, 'subscribe': 'subscribe'}
+            )
+        )
+
+    @client_user('student01', 'student01')
+    def test_subscription_fail(self):
+        """
+        Test failed on permission basis subscription.
+        """
+        qset_id = 10  # Qset 4-1 from the mockups
+        # Subscribe current user to the selected qset
+        response = self.change_qset_subscription(self, qset_id, 'subscribe').json()
+        self.assertEqual(response['result'], 'fail')
+        self.assertEqual(
+            response['url'],
+            reverse(
+                'askup:qset_subscription',
+                kwargs={'qset_id': qset_id, 'subscribe': 'subscribe'}
+            )
+        )
+
+        # Unsubscribe current user to the selected qset
+        response = self.change_qset_subscription(self, qset_id, 'unsubscribe').json()
+        self.assertEqual(response['result'], 'fail')
+        self.assertEqual(
+            response['url'],
+            reverse(
+                'askup:qset_subscription',
+                kwargs={'qset_id': qset_id, 'subscribe': 'unsubscribe'}
+            )
+        )
+
+
+class TestSubscriptionsMailing(TestCase):
+    """
+    Testing the subject subscription mailing functionality.
+    """
+
+    fixtures = ['groups', 'mockup_data']
+
+    def setUp(self):
+        """
+        Set up the test conditions.
+        """
+        settings.DEBUG = False
```

Comment:
Just curious -- is this necessary, cf https://docs.djangoproject.com/en/1.11/topics/testing/overview/#other-test-conditions?

Reply:
Thanks. I think those are redundant.

Did include all your suggestions into the next pull request.

Reply:
Thanks. I think those are redundant.

Did include all your suggestions into the next pull request.